sklearn/linear_model/_stochastic_gradient.py

```diff
@@ -284,7 +288,21 @@ def _make_validation_split(self, y):
         cv = splitter_type(
             test_size=self.validation_fraction, random_state=self.random_state
         )
-        idx_train, idx_val = next(cv.split(np.zeros(shape=(y.shape[0], 1)), y))
+
+        idx_non_zero = np.arange(n_samples)[sample_mask]
+        idx_train_, idx_val_ = next(cv.split(np.zeros(shape=(y.shape[0], 1)), y))
+
+        # make sure the sample weights for validation set is non-zero
+        idx_train = np.intersect1d(idx_non_zero, idx_train_)
+        idx_val = np.intersect1d(idx_non_zero, idx_val_)
+
+        cnt_zero_weight_val = len(idx_val_) - len(idx_val)
+        if cnt_zero_weight_val > 0:
+            warnings.warn(
+                f"There are {cnt_zero_weight_val} samples with zero sample weight in"
+                " the validation set, consider using a different random state."
```

Comment:
I think having some zero sample weights in the validation set is okay. I think the original issue was because all the samples weights were zero.

Reply:
Hi @thomasjpfan , the warning is added because we're changing the default bahaviour as you suggested in the previous comment
> Given that, I prefer to error or warn when this happens and suggest changing the random_state.

Reply:
In https://github.com/scikit-learn/scikit-learn/pull/23275#discussion_r911190871, I mentioned:

> As long as not all of them have zero weight, then there will still be a score.

To be clear, I meant to warn when all of the weights are zero. In other words:

```python
if not np.any(sample_mask[idx_val]):
    warnings.warn(...)
```

Reply:
Hi @thomasjpfan , I've updated the code to raise an error instead of warning because a `ZeroDivisionError` will eventually be raised when the sample weights for validation set are all 0, so I think it's useful to raise an error with a more informative error message beforehand.